### PR TITLE
[JUJU-1852] - Feature/sidecar open port

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR implements `open-port`, `close-port`, `opened-ports` for k8s sidecar applications.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
juju exec --unit prometheus-k8s/0 -- open-port 3000

juju exec --unit prometheus-k8s/0 -- open-port 3001

juju exec --unit prometheus-k8s/0 -- open-port 3002

juju exec --unit prometheus-k8s/0 -- open-port 3003

juju exec --unit prometheus-k8s/0 -- open-port 3004

juju exec --unit prometheus-k8s/0 -- opened-ports

3000/tcp
3001/tcp
3002/tcp
3003/tcp
3004/tcp

mkubectl -nt1 get service/prometheus-k8s -o json | jq '.spec.ports'
[
  {
    "name": "juju-3000-tcp",
    "port": 3000,
    "protocol": "TCP",
    "targetPort": 3000
  },
  {
    "name": "juju-3001-tcp",
    "port": 3001,
    "protocol": "TCP",
    "targetPort": 3001
  },
  {
    "name": "juju-3002-tcp",
    "port": 3002,
    "protocol": "TCP",
    "targetPort": 3002
  },
  {
    "name": "juju-3003-tcp",
    "port": 3003,
    "protocol": "TCP",
    "targetPort": 3003
  },
  {
    "name": "juju-3004-tcp",
    "port": 3004,
    "protocol": "TCP",
    "targetPort": 3004
  },
  {
    "name": "prometheus-k8s",
    "port": 9090,
    "protocol": "TCP",
    "targetPort": 9090
  }
]

juju exec --unit prometheus-k8s/0 -- close-port 3003

juju exec --unit prometheus-k8s/0 -- close-port 3004

juju exec --unit prometheus-k8s/0 -- opened-ports

3000/tcp
3001/tcp
3002/tcp

mkubectl -nt1 get service/prometheus-k8s -o json | jq '.spec.ports'
[
  {
    "name": "juju-3000-tcp",
    "port": 3000,
    "protocol": "TCP",
    "targetPort": 3000
  },
  {
    "name": "juju-3001-tcp",
    "port": 3001,
    "protocol": "TCP",
    "targetPort": 3001
  },
  {
    "name": "juju-3002-tcp",
    "port": 3002,
    "protocol": "TCP",
    "targetPort": 3002
  },
 {
    "name": "prometheus-k8s",
    "port": 9090,
    "protocol": "TCP",
    "targetPort": 9090
  }
]

juju exec --unit prometheus-k8s/0 -- open-port 3001-3003/tcp

ERROR port ranges are not supported for CAAS, please specify a single port
ERROR the following task failed:
 - id "16" with return code 1

use 'juju show-task' to inspect the failure

```

## Documentation changes

No

## Bug reference

No

### CODE REVIEW

File: worker/uniter/runner/context/context.go [hunk 1]
```diff
@@ -1408,14 +1408,21 @@ func (ctx *HookContext) doFlush(process string) error {
 		b.UpdateRelationUnitSettings(rctx.RelationTag().String(), unitSettings, appSettings)
 	}
 
-	for endpointName, portRanges := range ctx.portRangeChanges.pendingOpenRanges {
-		for _, pr := range portRanges {
-			b.OpenPortRange(endpointName, pr)
+	if len(ctx.portRangeChanges.pendingOpenRanges)+len(ctx.portRangeChanges.pendingCloseRanges) > 0 {
+		// Open/Close port can be done on leaders only for CAAS model.
+		if err := ctx.caasLeaderShipCheck(); err != nil {
```
Comment (hpidcock) [inline worker/uniter/runner/context/context.go:1413]: `LeaderShip` => `Leadership`

File: worker/uniter/runner/context/context.go [hunk 2]
```diff
@@ -1536,6 +1543,20 @@ func (ctx *HookContext) doFlush(process string) error {
 	return nil
 }
 
+func (ctx *HookContext) caasLeaderShipCheck() error {
```
Comment (hpidcock) [inline worker/uniter/runner/context/context.go:1546]: `LeaderShip` => `Leadership`

File: worker/uniter/runner/context/ports.go [hunk 3]
```diff
@@ -179,11 +199,15 @@ func (r *portRangeChangeRecorder) checkForConflict(incomingEndpoint string, inco
 // OpenedUnitPortRanges returns the set of port ranges currently open by the
 // current unit grouped by endpoint.
 func (r *portRangeChangeRecorder) OpenedUnitPortRanges() network.GroupedPortRanges {
-	return r.machinePortRanges[r.unitTag]
+	// return r.machinePortRanges[r.unitTag]
```
Comment (hpidcock) [inline worker/uniter/runner/context/ports.go:202]: remove comment

**Review by hpidcock — APPROVED** [review]
Wonderful. Few nitpicks, otherwise LGTM. We will probably need documentation updated somewhere. Do we also need any changes to the operator framework?

@tmihoc @benhoyt

**Comment by ycliuhw** [comment]
> Wonderful. Few nitpicks, otherwise LGTM. We will probably need documentation updated somewhere. Do we also need any changes to the operator framework?
> 
> @tmihoc @benhoyt

We don't have feature changes on the open/close port hook command here.
But for k8s sidecar, we only support a single port but no port range.
So no change need to be done on operator framework side.

**Pull request merged**